### PR TITLE
Delegation

### PR DESCRIPTION
**Pull request by andorp**
Squashed version of #615

### CODE REVIEW

File: src/Persistent/Idiom.hs [hunk 1]
```diff
@@ -202,3 +205,57 @@ ideaVoteSupport = ideaVoteSupportByAbsDiff
...
+                <$> forM path (scopeDelegatees user)
+        allNewDelegatees <- Set.unions
+            <$> forM (Set.toList oneStepNewDelegatees)
+                     (voters path (oneStepNewDelegatees `Set.union` discovered))
+        pure (discovered `Set.union` allNewDelegatees)
+
+scopeDelegatees :: AUID User -> DScope -> Query [Delegation]
+scopeDelegatees uid scope =
+    filter ((&&) <$> ((uid ==) . view delegationTo)
+                 <*> ((scope ==) . view delegationContext))
```
Comment (andorp) [inline src/Persistent/Idiom.hs]: The `delegationContext` should also be renamed.

File: src/Frontend/Page/Delegation.hs [hunk 2]
```diff
@@ -14,31 +14,89 @@ where
...
 -- | 12. Delegate vote
-data PageDelegateVote = PageDelegateVote
+data PageDelegateVote = PageDelegateVote DScopeFull [User]
   deriving (Eq, Show, Read)
 
-instance ToHtml PageDelegateVote where  -- FIXME: remove bogus instance.
-    toHtmlRaw = toHtml
-    toHtml p = semanticDiv p "PageDelegateVote"
-
 instance Page PageDelegateVote where isAuthorized = adminPage -- FIXME who needs to see this
```
Comment (andorp) [inline src/Frontend/Page/Delegation.hs]: `PageDelegateVote` is a `userPage`.

File: src/Frontend/Page/Delegation.hs [hunk 3]
```diff
@@ -14,31 +14,89 @@ where
...
-    formPage _ _ _ = pure ()
+newtype PageDelegationVotePayload = PageDelegationVotePayload
+    { unPageDelegationVotePayload :: AUID User }
+  deriving (Eq, Show, Read)
+
+instance FormPage PageDelegateVote where
+    type FormPagePayload PageDelegateVote = PageDelegationVotePayload
+
+    formAction (PageDelegateVote scope _users) = case scope of
+        DScopeGlobalFull          -> U.Broken
```
Comment (andorp) [inline src/Frontend/Page/Delegation.hs]: No delegation page is rendered in this situation. Should we remove this explicitly? @fisx

**Comment by fisx** [comment]
LGTM.  this is a big piece of code!

there are a few TODOs left in the code, but i will merge when green and we should deal with them later.

**Pull request merged**